Introduce `FeatureMesh`

gchoqueux opened the pull request:
## Description
Introduce `FeatureMesh` to handle the feature converted to mesh.
`FeatureMesh` could be it can dynamically change its geographic projection system.
This method avoids to use `proj4`, which is more accurate less efficient.
It's still possible to use `proj4` with an `FeatureGeometryLayer` option.

 `FeatureMesh`  is used with `FeatureGeometryLayer`. 
Now, the 3d objects of `FeatureGeometryLayer` are added to `FeatureGeometryLayer.object3d`.

## Motivation and Context
Avoid `proj4` increases performance.
The `feature processing` is more simple and efficient.
`FeatureMesh` allows to display the tiled vector in 3d without any geographical transformation (directely from Protobufs raw data )

## Screenshots
                                  
![image](https://user-images.githubusercontent.com/11291849/160877104-540dda93-4dca-4ce7-a223-6b7915f03c7a.png)

jailln commented on examples/vector_tile_3d_mesh.html:
```diff
@@ -0,0 +1,104 @@
+                addLabelLayer: true,
+            });
+
+            var mvtBuildingLayer = new itowns.FeatureGeometryLayer('WFS Building',{
```
MVT Buildings instead of WFS buildings ?

gchoqueux replied:
yes

jailln commented on src/Layer/FeatureGeometryLayer.js:
```diff
@@ -22,22 +22,37 @@ class FeatureGeometryLayer extends GeometryLayer {
-        super(id, config.object3d || new Group(), config);
+        super(id, options.object3d || new Group(), options);
         this.isFeatureGeometryLayer = true;
+        this.useProj4 = options.useProj4 ?? false;
```
Shouldn't it be true by default ? So it is by default more precise geographically but you can use an option to increase performances ?

jailln commented:
Nice feature! Could you quantify how faster it is to not use proj4 on an example dataset ? And to estimate what is the geometric loss ?

gchoqueux commented:
> Nice feature! Could you quantify how faster it is to not use proj4 on an example dataset ? And to estimate what is the geometric loss ?

the `parse` is twice as fast. In wfs example the error is inferior to 0.01 mm.

gchoqueux commented:
I will add a test to calculate and verify the geometric loss

jailln commented:
> > Nice feature! Could you quantify how faster it is to not use proj4 on an example dataset ? And to estimate what is the geometric loss ?
> 
> the `parse` is twice as fast. In wfs example the error is inferior to 0.01 mm.

Cool! Could you add it in the doc please ? So one knows what it implies to use one option or the other.

mgermerie commented on src/Process/FeatureProcessing.js:
```diff
@@ -20,15 +19,12 @@ export default {
             node.layerUpdateState[layer.id] = new LayerUpdateState();
         } else if (!node.layerUpdateState[layer.id].canTryUpdate()) {
+            // toggle visibility features
+            const features = node.link.filter(n => n.layer && (n.layer.id == layer.id));
```
Would it be possible to replace callback with `n => n.layer?.id === layer.id ` ?

gchoqueux replied:
yes

mgermerie commented on src/Parser/VectorTileParser.js:
```diff
@@ -62,17 +62,15 @@ function vtFeatureToFeatureGeometry(vtFeature, feature, classify = false) {
                 firstPoint.set(x, y);
                 lastPoint.set(x, y);
-            } else if (classify && count > 1) {
+            } else if (count > 1) {
```
Maybe it could be in a separate PR/commit ?

gchoqueux replied:
yes

mgermerie commented on src/Process/FeatureProcessing.js:
```diff
@@ -54,31 +50,27 @@ export default {
+        return context.scheduler.execute(command).then((meshes) => {
+            node.layerUpdateState[layer.id].noMoreUpdatePossible();
+
+            meshes.forEach((featureNode) => {
```
Could you rename `featureNode` to `featureMesh` ?

gchoqueux replied:
yes

mgermerie commented on src/Process/FeatureProcessing.js:
```diff
@@ -54,31 +50,27 @@ export default {
+            node.layerUpdateState[layer.id].noMoreUpdatePossible();
+
+            meshes.forEach((featureNode) => {
+                if (featureNode) {
```
Is it necessary to do this check ? I mean : can the commend return an array containing an `undefined` value if the matching extent has no associated data ?

gchoqueux replied:
indeed, there seems to be no case where the `featureMesh` is undefined

gchoqueux replied:
but it's necessary to add PR to solve this issue

mgermerie commented on src/Core/Feature.js:
```diff
@@ -145,7 +145,9 @@ export class FeatureGeometry {
-        _setGeometryValues(this, feature, coordOut.x, coordOut.y, coordOut.z, coordOut.geodesicNormal);
+        const normal = coordIn.crs == feature.crs ? defaultNormal : coordOut.geodesicNormal;
+
+        _setGeometryValues(this, feature, coordOut.x, coordOut.y, coordOut.z, normal);
```
The `normal` array could be simplified if all normal are worth `defaultNormal`.

gchoqueux replied:
I add a new PR to resolve this issue

mgermerie commented on src/Process/FeatureProcessing.js:
```diff
@@ -54,31 +50,27 @@ export default {
+                if (featureNode) {
+                    featureNode.as(context.view.referenceCrs);
+
+                    if (layer.onMeshCreated) {
```
On another subject than this PR, the `onMeshCreated` property has no documentation to be found anywhere.

gchoqueux replied:
added

mgermerie commented on src/Process/FeatureProcessing.js:
```diff
@@ -54,31 +50,27 @@ export default {
+                    featureNode.as(context.view.referenceCrs);
+
+                    if (layer.onMeshCreated) {
+                        layer.onMeshCreated(featureNode, context);
```
The `context` parameter is not used in any examples or source code. It can be a useful addition, but would maybe need an example and docs.

gchoqueux replied:
in another PR

mgermerie commented on src/Process/FeatureProcessing.js:
```diff
@@ -54,31 +50,27 @@ export default {
-                if (layer.onMeshCreated) {
-                    layer.onMeshCreated(result);
-                }
-                node.layerUpdateState[layer.id].success();
```
Don't we need to specify it anymore ?

gchoqueux replied:
is replaced by `noMoreUpdatePossible()`

mgermerie commented on src/Process/FeatureProcessing.js:
```diff
@@ -54,31 +50,27 @@ export default {
-                node.add(group.add(result));
-                group.updateMatrixWorld(true);
-            } else {
-                node.layerUpdateState[layer.id].failure(1, true);
```
Same question as for the `success` : we don't manage update state in case fetched data is empty.

mgermerie commented on src/Converter/Feature2Mesh.js:
```diff
@@ -3,12 +3,72 @@ import Earcut from 'earcut';
+            coord.setFromVector3(this.meshesCollection.position);
+
+            // get method to calculate orientation
+            const crsWGS84 = 'EPSG:4326';
```
This could be defined outside the method.

mgermerie commented on src/Process/FeatureProcessing.js:
```diff
@@ -54,31 +50,27 @@ export default {
-                group.layer = layer;
-                group.matrixWorld.copy(node.matrixWorld).invert();
-                group.matrixWorld.decompose(group.position, group.quaternion, group.scale);
-                group.position.z += node.geoidHeight;
```
The geoid height is no longer managed. You could add 
```js
featureNode.place.position.z += node.geoidHeight;
featureNode.updateMatrixWorld();
```
just after `featureNode.as(context.view.referenceCrs);` on line 58.

There must be a cleaner way of doing this (to avoid calling `updateMatrixWorld` both in `FeatureMesh.as` method and just after).

mgermerie commented on src/Converter/Feature2Mesh.js:
```diff
@@ -29,14 +30,16 @@ class FeatureMesh extends THREE.Group {
         this.meshesCollection.updateMatrix();
 
         this.crs = collection.crs;
+        this.#currentCrs = this.crs;
```
In my opinion, it would make more sense to rename `this.crs` to something like `this.collectionCrs` and `this.#currentCrs` to `this.#crs`. WDYT ?

Also, I don't see the the current `this.crs` property being used outside of `FeatureMesh` class. Could it be set private ?

gchoqueux replied:
I simplified the code

mgermerie commented on examples/vector_tile_3d_mesh.html:
```diff
@@ -0,0 +1,157 @@
+
+        <script type="text/javascript">
+
+             var promises = [];
```
I forgot this unused line. Could you remove it ?

gchoqueux replied:
solved

mgermerie commented:
There is another issue with geoid height : 
If i display a `GeoidLayer` along with a `GeometryLayer`, toggling the `GeoidLayer` visibility won't translate `GeometryLayer` meshes, as can be seen in both screenshots bellow.

Geoid visibility ON :
![geoid-on](https://user-images.githubusercontent.com/73115044/161028061-b26bea4c-1959-454a-81f7-30dc1fc2e484.png)

Geoid visibility OFF : 
![geoid-off](https://user-images.githubusercontent.com/73115044/161028197-edb75117-3949-44b4-8924-0c337d2b0de1.png)

This behavior does not appears on master branch.

gchoqueux commented:
The geoid issue is resolved
But there is an geoid issue on master, i will fix it after the PR merging

gchoqueux committed:
test(Feature2Mesh): add test to calculate the difference with and without proj4.

gchoqueux committed:
example(FeatureGeometryLayer): add vector tile to 3d object.

mgermerie commented on src/Process/FeatureProcessing.js:
```diff
@@ -20,15 +19,12 @@ export default {
             node.layerUpdateState[layer.id] = new LayerUpdateState();
         } else if (!node.layerUpdateState[layer.id].canTryUpdate()) {
+            // toggle visibility features
+            const features = node.link.filter(n => n.layer?.id == layer.id);
```
It is a detail, but it might be more relevant to rename `n` parameter to `feature` or `f`, since `node.link` contains `FeatureMeshes`.

mgermerie commented on src/Layer/FeatureGeometryLayer.js:
```diff
@@ -22,22 +22,45 @@ class FeatureGeometryLayer extends GeometryLayer {
+     * @param {THREE.Object3D} [options.object3d=new THREE.Group()] root object3d layer.
+     * @param {function} [options.onMeshCreated] this callback is called when the mesh is created. The callback parameters are the
+     * `mesh` and the `context`.
+     * @param {boolean} [options.accurate=TRUE] If `accurate` is `true` the data are reprojected with maximum geographical accurate.
```
Could you replace by :
```
* @param {boolean} [options.accurate=TRUE] If `accurate` is `true`, data are re-projected with maximum geographical accuracy.
* With `true`, `proj4` is used to transform data source.
*
* If `accurate` is `false`, re-projecting is faster but less accurate.
* With `false`, an affine transformation is used to transform data source.
* This method is an approximation. The error increases with the extent
* dimension of the object or queries.

* For example : 
...
```

gchoqueux committed:
feature(FeatureGeometryLayer): introduce FeatureMesh, they are added to layer.object3d.

FeatureMesh could be it can dynamically change its geographic projection system.
